### PR TITLE
feat(queue): Make instance status externally settable

### PR DESCRIPTION
**Pull request by lwander**
Note this only works when the instance is not managed by eureka.

### CODE REVIEW

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy [hunk 1]
```diff
@@ -232,4 +238,25 @@ class OperationsController {
...
+      return
+    }
+
+    if (!discoveryStatusPoller instanceof  NoDiscoveryApplicationStatusPublisher) {
+      throw new IllegalStateException("Discovery status cannot be overwritten - there is an existing discovery client managing orca")
+    }
+
+    NoDiscoveryApplicationStatusPublisher noDiscoveryApplicationStatusPublisher = (NoDiscoveryApplicationStatusPublisher) discoveryStatusPoller;
+
+    boolean enabled = enabledWrapper.get("enabled")
```
Comment (robfletcher) [inline orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy]: Why not just a boolean param?

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy [hunk 2]
```diff
@@ -232,4 +238,25 @@ class OperationsController {
...
+  @RequestMapping(value = "/instance/enabled", method = RequestMethod.POST)
+  void setInstanceStatus(@RequestBody Map<String, Boolean> enabledWrapper) {
+    if (!discoveryStatusPoller) {
+      log.warn("No configured discovery status poller")
+      return
+    }
+
+    if (!discoveryStatusPoller instanceof  NoDiscoveryApplicationStatusPublisher) {
+      throw new IllegalStateException("Discovery status cannot be overwritten - there is an existing discovery client managing orca")
+    }
```
Comment (robfletcher) [inline orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy]: what HTTP status will this result in? Would make sense to return 501, maybe?

Comment (lwander) [inline orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy]: 500 - 501 seems reasonable

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy (hunk 1 above)
Comment (lwander) [inline orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy]: A boolean param alone wasn't being deserialized correctly/I don't know what content type makes sense there - do you have an example of this working?

Comment (robfletcher) [inline orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy]: oh I see, maybe I've done it with a query string before. Doesn't really matter

**Comment by robzienert** [comment]
👍

**Comment by lwander** [comment]
Sorry for the wait, was OOO, will set the error type to 501 now

**Comment by lwander** [comment]
Actually, 405 (method not allowed) kind of makes more sense since we're rejecting the user's attempted state change. Any objections?

**Comment by lwander** [comment]
It is a POST - and having a special error code makes it easier for Halyard as a client to figure out why it can't change discovery status. In this case the error means all is well, orca will deregister itself once disabled.

**Comment by lwander** [comment]
Oh you mean rejecting POST, PATCH, or PUT as in "don't use these", not, "don't use these like that"? Hmm then I should find something different.

**Comment by lwander** [comment]
Maybe 501 does make the most sense from the perspective of "at some point we can flip eureka status from orca"

**Comment by ajordens** [comment]
Is this `POST` going to get fronted by `gate`? ... if so I suspect the exception handling there is going to catch this and return it as a `500` anyhow.

Also 501 does have a generally intended meaning that the HTTP method you supplied (HTTP POST) is not implemented on the server.

If you don't want to differentiate by message, would returning an additional attribute in the error payload let Halyard do what it needs to do?

See the `HasAdditionalAttributes` interface that an exception can implement.

```
@ResponseStatus(HttpStatus.NOT_IMPLEMENTED)
 private static class DiscoveryUnchangeableException extends IllegalStateException implements HasAdditionalAttributes {
   Map<String, Object> additionalAttributes = new HashMap<>();

   DiscoveryUnchangeableException(String message, int errorCode) {
     super(message)
     additionalAttributes.put("errorCode", errorCode);
   }
  
   Map<String, Object> getAdditionalAttributes() { return additionalAttributes; }
```

Something like the above should work.

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy [hunk 3]
```diff
@@ -232,4 +236,32 @@ class OperationsController {
     def pipeline = orchestrationLauncher.start(json)
     [ref: "/tasks/${pipeline.id}".toString()]
   }
+
+  @RequestMapping(value = "/instance/enabled", method = RequestMethod.POST)
```
Comment (ajordens) [inline orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy]: Thinking to the future ... any value in rooting these sort of admin-like operations under a `/admin` path?

(I wouldn't be surprised if someday we support manipulating queue throttles via an admin API as well)

Might make it easier to specifically attach additional auth configurations (for example).

Comment (lwander) [inline orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy]: Oh I like that - I'll create a separate controller for that.

**Comment by lwander** [comment]
Halyard will be opening an ssh tunnel & talking directly to orca for this, so we don't have to worry about gate squashing the error.

I remember spring only copying over the error message & class when returning to a client - let me check that the additional attributes are preserved. If so, that sounds like a good way to deal with this.

**Comment by lwander** [comment]
Moved into a new admin controller, and took the `HasAdditionalAttributes` approach

**Pull request merged**